Expand the value of math and reduce expressions in annotations and labels

grobinson-grafana opened the pull request:
**What this PR does / why we need it**:

This commit makes it possible to use the values of reduce and math expressions in annotations and labels via their RefIDs. An example of such an annotation could be "The value of A is `{{ $values.A }}"`.

**Which issue(s) this PR fixes**:

Fixes #36020 

**Special notes for your reviewer**:

grobinson-grafana committed:
Expand the value of math and reduce expressions in annotations and labels

This commit makes it possible to use the values of reduce and math
expressions in annotations and labels via their RefIDs. An example
of such an annotation could be "The value of A is {{ $values.A }}".

gerobinson commented on pkg/services/ngalert/eval/eval.go:
```diff
@@ -72,6 +72,11 @@ type Result struct {
+	// Values contains the RefID and value of reduce and math expressions.
+	// It does not contain values for classic conditions as the values
+	// in classic conditions do not have a RefID.
+	Values map[string]*float64
```
@davidmparrott could this be problematic for how we cache evaluations? I think its all cached under the Org ID and the Rule ID, neither of which would be affected given this change.

gerobinson replied:
@davidmparrott could this be problematic for how we cache evaluations? I think its cached on the Org ID and the Rule ID, neither of which would be affected given this change.

kylebrandt commented on pkg/services/ngalert/state/cache_test.go:
```diff
@@ -0,0 +1,42 @@
+package state
+
+import (
+	ptr "github.com/xorcare/pointer"
```
Oh I'm glad this is catching on :-)

kylebrandt commented:
I think maybe we need this under something other than `$labels`. This is because there could be a collision with the RefID name inputted and labels returned from the data. E.g. if you named a refId `host` (note: if you hover over the RefID in the UI, you can edit it) and there was also a `host` label with the alert.

gerobinson commented:
> I think maybe we need this under something other than `$labels`. This is because there could be a collision with the RefID name inputted and labels returned from the data. E.g. if you named a refId `host` (note: if you hover over the RefID in the UI, you can edit it) and there was also a `host` label with the alert.

I agree that this could be an issue. Two options I can think of:

1. Put a suffix at the start of each RefID such that we have `$labels.refA`
2. Change the template to support two variables: `$labels` and `$refs`

kylebrandt commented:
> > I think maybe we need this under something other than `$labels`. This is because there could be a collision with the RefID name inputted and labels returned from the data. E.g. if you named a refId `host` (note: if you hover over the RefID in the UI, you can edit it) and there was also a `host` label with the alert.
> 
> I agree that this could be an issue. Two options I can think of:
> 
> 1. Put a suffix at the start of each RefID such that we have `$labels.refA`
> 2. Change the template to support two variables: `$labels` and `$refs`

I think it should be option 2 (maybe a different name than `$refs` ... but I don't have a better one). I think the different object is important because the information is not a label, so that could be confusing to users. Also adding a prefix only makes the collision less likely, as one could still have a label with that prefix.

gerobinson commented:
> > > I think maybe we need this under something other than `$labels`. This is because there could be a collision with the RefID name inputted and labels returned from the data. E.g. if you named a refId `host` (note: if you hover over the RefID in the UI, you can edit it) and there was also a `host` label with the alert.
> > 
> > 
> > I agree that this could be an issue. Two options I can think of:
> > 
> > 1. Put a suffix at the start of each RefID such that we have `$labels.refA`
> > 2. Change the template to support two variables: `$labels` and `$refs`
> 
> I think it should be option 2 (maybe a different name than `$refs` ... but I don't have a better one). I think the different object is important because the information is not a label, so that could be confusing to users. Also adding a prefix only makes the collision less likely, as one could still have a label with that prefix.

I propose `$values.A` etc and have updated the pull request to show those changes.

kylebrandt commented:
I like `.Values`. Seems to work on the dashboard for me  (See images below), still have to test notifications

I'm wondering if we don't want a structure here the is more map[string]NumberValueCapture (or some struct), that way in templating, people can loop over things and pull out all the information that is included in ValueString if they want - However it also might mean slightly more complicated templating, e.g. `{{ .Values.A.Value }}` - wonder what others think since this would be hard to change going forward.

Whichever we go with, should update https://github.com/grafana/grafana/blob/main/docs/sources/alerting/unified-alerting/message-templating/template-data.md to contain this (and maybe `.ValueString` if you don't mind since I missed that last time, and will be needed for CC based alerts).

![image](https://user-images.githubusercontent.com/1692624/125096309-039f6180-e0a3-11eb-81ee-46442ec56f1d.png)

![image](https://user-images.githubusercontent.com/1692624/125096345-0ef28d00-e0a3-11eb-9f65-99617c5ecb86.png)

kylebrandt commented on pkg/services/ngalert/state/cache.go:
```diff
@@ -224,6 +223,25 @@ func (c *cache) recordMetrics() {
+		labels[k] = v
+	}
+	values := make(map[string]string)
+	for k, v := range result.Values {
```
It might be better if this is an actual float64 (or if we go with the struct I mentioned earlier, the property would be float), I am think this might be the case so people can do number formatting with the template built-in `printf`

gerobinson replied:
I think it would need to be `*float64` since the value can be absent? I'm not sure I want to use `0` because `0` is a valid value and in the value string we use `"null"` in places of absent values. If we use `*float64` do we need to be concerned about templates dereferencing nil?

gerobinson replied:
I think it would need to be `*float64` since the value can be absent?

I don't think I want to use `0` because `0` is a valid metric value. In the value string we use `"null"` in places of absent values. If we use `*float64` do we need to be concerned about templates dereferencing nil?

gerobinson replied:
I think it would need to be `*float64` since the value can be absent? I don't think I want to use `0` because `0` is a valid metric value. In the value string we use `"null"` in places of absent values. If we use `*float64` do we need to be concerned about templates dereferencing nil?

kylebrandt replied:
Yes pointer seems right. It can be nil, because {{ pipeline }} is calling printf

Example to show it doesn't panic: https://play.golang.org/p/_LmX3iciRG9

kylebrandt replied:
Oh wait no I'm wrong, will print address the value... Wonder how prom does the number formating ...

kylebrandt replied:
Looks like `humanzie` https://prometheus.io/docs/prometheus/latest/configuration/template_reference/#numbers ... not sure what is best here https://github.com/JensErat/prometheus/blob/master/template/template.go#L167

Float64 non-pointer is maybe fine if we create the container, as I expect that the entry won't exist if there wasn't a number for it so the containing object wouldn't exist?

gerobinson commented:
> I'm wondering if we don't want a structure here the is more map[string]NumberValueCapture (or some struct), that way in templating, people can loop over things and pull out all the information that is included in ValueString if they want - However it also might mean slightly more complicated templating, e.g. `{{ .Values.A.Value }}` - wonder what others think since this would be hard to change going forward.

Yes I think we could. Could we add a `String() string` method to the struct for `{{ .Values.A }}`? I'm not sure if `text/template` uses the `fmt.Stringer` interface?
 
> Whichever we go with, should update https://github.com/grafana/grafana/blob/main/docs/sources/alerting/unified-alerting/message-templating/template-data.md to contain this (and maybe `.ValueString` if you don't mind since I missed that last time, and will be needed for CC based alerts).

I can do that! :)

mkopinsky commented:
re syntax - (apologies since I'm hugely missing context, you can blame @davidmparrott for inviting slack randos into your PR reviews if you'd like 😉): 

Do all of the proposed syntaxes allow spaces in expression/query names? Currently I can do something like `${free connections}` in a Math expression, and it would be nice if you could do the same in templates.

kylebrandt commented:
> Do all of the proposed syntaxes allow spaces in expression/query names? Currently I can do something like `${free connections}` in a Math expression, and it would be nice if you could do the same in templates.

Yes, it is my screenshots above (but hard to see and it got wrapped). Syntax is `{{ index .Values "is GT" }}` from my example above, where "is GT" is the renamed RefId / SSE variable.

gerobinson commented:
> I'm wondering if we don't want a structure here the is more map[string]NumberValueCapture (or some struct), that way in templating, people can loop over things and pull out all the information that is included in ValueString if they want - However it also might mean slightly more complicated templating, e.g. `{{ .Values.A.Value }}` - wonder what others think since this would be hard to change going forward.

We can do that either with `eval.NumberValueCapture` or having a similar struct that is declared in the `state` package. I don't think it makes much difference at this present time.

I can also confirm that `text/template` uses the fmt `Stringer` interface so we can create a `String() string` method on `eval.NumberValueCapture` (or whatever struct we use) and would mean we can still write `{{ .Values.A }}` and get nice output.

grobinson-grafana committed:
Expand both the labels and the value of reduce and math expressions

This commit makes it possible to expand both the labels and the
value of reduce and math expressions in annotations and labels.

It uses the Stringer interface to ensure that "{{ $values.A }}"
still prints the value in decimal format while also making the
labels for each RefID available with "{{ $values.A.Labels }}" and
the float64 value with "{{ $values.A.Value }}"

gerobinson commented:
I have pushed a commit with the following changes:

- `{{ $values.A }}` prints the decimal value for the RefID A or `null` if the value is absent
- `{{ $values.A.Labels }}` prints the labels for the RefID A
- `{{ $values.A.Value }}` prints the float64 value for the RefID A

gerobinson commented:
I have pushed a commit with the following changes:

- `{{ $values.A }}` prints the decimal value for the RefID A or `null` if the value is absent
- `{{ $values.A.Labels }}` prints the labels for the RefID A
- `{{ $values.A.Value }}` prints the float64 value for the RefID A

gerobinson commented:
I have pushed a commit with the following changes:

- `{{ $values.A }}` prints the decimal value for the RefID A or `null` if the value is absent
- `{{ $values.A.Labels }}` prints the labels for the RefID A
- `{{ $values.A.Value }}` prints the float64 value for the RefID A

gerobinson commented:
I have pushed a commit with the following changes:

- `{{ $values.A }}` prints the decimal value for the RefID A or `null` if the value is absent
- `{{ $values.A.Labels }}` prints the labels for the RefID A
- `{{ $values.A.Value }}` prints the float64 value for the RefID A

gerobinson commented:
I have pushed a commit with the following changes:

- `{{ $values.A }}` prints the decimal value for the RefID A or `null` if the value is absent
- `{{ $values.A.Labels }}` prints the labels for the RefID A
- `{{ $values.A.Value }}` prints the float64 value for the RefID A

gerobinson commented:
I have pushed a commit with the following changes:

- `{{ $values.A }}` prints the decimal value for the RefID A or `null` if the value is absent
- `{{ $values.A.Labels }}` prints the labels for the RefID A
- `{{ $values.A.Value }}` prints the float64 value for the RefID A

gotjosh commented on pkg/services/ngalert/eval/extract_md.go:
```diff
@@ -65,3 +65,24 @@ func extractEvalString(frame *data.Frame) (s string) {
+// no results in the frame.
+func extractValues(frame *data.Frame) map[string]NumberValueCapture {
+	if frame == nil {
+		return nil
```
Is `nil` the same as an empty map for all cases? I'm not very familiar with how go templates work but if we ended up doing `values.A` in the template wouldn't this panic in this case?

kylebrandt replied:
I thought it would panic, but I can't seem to get it to https://play.golang.org/p/cQeGYhhHesL

gotjosh replied:
TIL

gerobinson replied:
The check for that is here https://github.com/golang/go/blob/master/src/text/template/exec.go#L976

gotjosh commented on pkg/services/ngalert/eval/extract_md.go:
```diff
@@ -65,3 +65,24 @@ func extractEvalString(frame *data.Frame) (s string) {
+		return nil
+	}
+	if caps, ok := frame.Meta.Custom.([]NumberValueCapture); ok {
+		v := make(map[string]NumberValueCapture)
```
```suggestion
		v := make(map[string]NumberValueCapture, len(caps))
```

gotjosh commented on pkg/services/ngalert/state/cache.go:
```diff
@@ -35,12 +36,13 @@ func (c *cache) getOrCreate(alertRule *ngModels.AlertRule, result eval.Result) *
 	defer c.mtxStates.Unlock()
 
-	templateData := make(map[string]string, len(result.Instance)+3)
+	labels := make(map[string]string)
```
```suggestion
	labels := result.Instance.Copy()
```
nit: `Labels` have a method to copy themselves.

gotjosh commented on pkg/services/ngalert/state/cache.go:
```diff
@@ -104,13 +106,28 @@ func (c *cache) expandRuleLabelsAndAnnotations(alertRule *ngModels.AlertRule, da
+// and labels template
+type templateCaptureValue struct {
+	Labels map[string]string
+	Value  *float64
```
Mostly for the benefit of my understanding: Are they always float, from previous conversations with @kylebrandt I thought there could be other types too? I guess they are given this is a `NumberValueCapture` as its name implies but would like to clear my understanding.

gerobinson replied:
`NumberValueCapture` must contain `float64`. I couldn't see values other than float64 being read in the `eval` package for either `classic.EvalMatch` or `NumberValueCapture`, but I could have missed something!

kylebrandt replied:
Always *float64 coming from SSE. It will take in other number types from data source queries (e.g. uInts etc) but they are converted to floats on ingestion by SSE.

gotjosh replied:
Thanks!

gotjosh commented on pkg/services/ngalert/state/cache.go:
```diff
@@ -104,13 +106,28 @@ func (c *cache) expandRuleLabelsAndAnnotations(alertRule *ngModels.AlertRule, da
 
-func expandTemplate(name, text string, data map[string]string) (result string, resultErr error) {
+// templateCaptureValue represents each value in .Values in the annotations
+// and labels template
```
```suggestion
// and labels template.
```
nit

gotjosh commented on pkg/services/ngalert/state/cache.go:
```diff
@@ -128,12 +145,22 @@ func expandTemplate(name, text string, data map[string]string) (result string, r
-	if err != nil {
+		Labels: labels,
+		Values: func() map[string]templateCaptureValue {
+			m := make(map[string]templateCaptureValue)
```
```suggestion
			m := make(map[string]templateCaptureValue, len(values))
```

gotjosh commented on pkg/services/ngalert/state/state.go:
```diff
@@ -28,6 +28,28 @@ type Evaluation struct {
+}
+
+// NewEvaluationValues returns the labels and values for each RefID in the capture.
+func NewEvaluationValues(m map[string]eval.NumberValueCapture) map[string]EvaluationValue {
```
Apologies if this is a dumb question, but why do we need this conversion? Couldn't we just use `map[string]eval.NumberValueCapture` directly? We're already converting these in the template to `templateCaptureValue`.

gerobinson replied:
I wanted to create a separation between the evaluation results and what we store in the cache as I think it will make it easier to make isolated changes to the code in future. It's not something I feel that strong about though and I'm more than open to changing it.

gotjosh replied:
If the only purpose is the separation of concerns - I'd be in favour of removing it, it's not doing much for us and the only way this is exposed to users is through the template, so we're covered in that area.

kylebrandt replied:
@gotjosh I think it might be good to have a different type here. Since we expose Go templates, I think it is important to have a very well defined interface. Else it can become possible for properties to get added to types, and this leaks into what can be accessed in a template.

gotjosh replied:
I 100% agree with that, but I thought that was:

```golang
// templateCaptureValue represents each value in .Values in the annotations
// and labels template.
type templateCaptureValue struct {
	Labels map[string]string
	Value  *float64
}
```

and not `EvaluationValue`, it seems like we're doing that twice if I'm reading this right.

gotjosh replied:
Had an offline discussion with @gerobinson and I'm happy to leave this as is.

gotjosh commented on pkg/services/ngalert/state/state.go:
```diff
@@ -28,6 +28,28 @@ type Evaluation struct {
+
+// NewEvaluationValues returns the labels and values for each RefID in the capture.
+func NewEvaluationValues(m map[string]eval.NumberValueCapture) map[string]EvaluationValue {
+	result := make(map[string]EvaluationValue)
```
```suggestion
	result := make(map[string]EvaluationValue, len(m))
```

Apologies if I'm coming across as pedantic with the hints, they're quite important in this area given this is touch by very often by every alert _instance_.

gotjosh reviewed (commented):
Looking good, great job @gerobinson! 👌 

I just have a few comments, mostly for the benefit of my understanding. After those, we should be ready to merge this.

I imagine that the documentation for this would come in a separate PR, right?

grobinson-grafana committed:
Fix feedback

grobinson-grafana committed:
Add documentation for template variables in annotations and labels

gotjosh commented on docs/sources/alerting/unified-alerting/alerting-rules/create-grafana-managed-rule.md:
```diff
@@ -103,6 +103,15 @@ Labels are key value pairs that categorize or identify an alert. Labels are  use
+/ Name    / Description     /
+/ ------- / --------------- /
+/ $labels / Labels contains the labels from the query or condition. For example, `{{ $labels.instance }}` and `{{ $labels.job }}`. /
+/ $values / Values contains the values of all reduce and math expressions that were evaluated for this alert rule. For example, `{{ $values.A }}`, `{{ $values.A.Labels }}` and `{{ $values.A.Value }}`. /
```
```suggestion
/ $values / Values contains the values of all reduce and math expressions that were evaluated for this alert rule. For example, `{{ $values.A }}`, `{{ $values.A.Labels }}` and `{{ $values.A.Value }}` where `A` is the `refID` of the expression. /
```

gotjosh reviewed (approved):
LGTM
